Automattic/woocommerce-payments
# Add a reference to the transpiled JS file in POT file comments

Fixes #1894

Currently the translations for ts/tsx files are not being exported to the JSON file when generating the language packs in translate.wordpress.com (only translations strings from js files are included).

The JSON files in the language pack contain all strings from the POT file grouped by `.js` filename (using the file reference for each string). This PR fixes the problem by adding additional file references to the generated (webpack transpiled) files for each string.

Example POT file:
```
#: client/additional-methods-setup/methods-selector/add-payment-methods-task.js:391
#: client/connect-account-page/modal/index.js:82
#: client/multi-currency-setup/tasks/store-settings-task/index.js:147
#: client/payment-methods/activation-modal.js:47
msgid "Continue"
msgstr ""
```
becomes:
```
#: dist/index.js:1
#: dist/settings.js:1
#: client/additional-methods-setup/methods-selector/add-payment-methods-task.js:391
#: client/connect-account-page/modal/index.js:82
#: client/multi-currency-setup/tasks/store-settings-task/index.js:147
#: client/payment-methods/activation-modal.js:47
msgid "Continue"
msgstr ""
```

More context in the slack thread: p1638277640394300-slack-CGGCLBN58

#### Changes proposed in this Pull Request

This PR extends the POT combination process to add a comment reference with the generated JS asset file (in the `dist` folder) for each javascript/typescript source file (in the `client` folder).

It also changes the `release.js` task to generate `hidden-source-maps` during production build, they are used to generate the final translations POT file and are not included in the zip package.
An alternative would be to use a Webpack plugin to generate the dependencies file ([like this](https://gist.github.com/diegocurbelo/b1905ab3e1cbd5c5549d9f34eca23098)), but using this plugin does not improve the build time over using hidden-source-maps:
- Develop branch:
`npm run build  113.43s user 7.20s system 274% cpu 43.928 total`
- This branch:
`npm run build  114.46s user 6.95s system 281% cpu 43.070 total`

The times for development builds are not affected, source-maps were already being generated for development builds

#### Testing instructions

1. Checkout `develop`
2. Run `npm run build`
3. Open the generated `languages/woocommerce-payments.pot` file
4. Search for `Risk level` and verify that there is only one reference file in the comments (a tsx file):
```
#: client/transactions/list/index.tsx:132
msgid "Risk level"
msgstr ""
```
5. Checkout this branch
6. Run `npm run build` again
7. Open the generated `languages/woocommerce-payments.pot` file
8. Search for `Risk level` again and verify that this time a reference file to the generated file was included:
```
#: dist/index.js:1
#: client/transactions/list/index.tsx:132
msgid "Risk level"
msgstr ""
```

#### Next steps

With this change the language packs will generate a JSON for each generated file (`dist/index.js`, `dist/settings.js`, etc), so and we can remove [wc-payments-translations-loader.php](https://github.com/Automattic/woocommerce-payments/blob/db27ff47f8d7f74b0d8c1fdef2a1ae57f9a8aab8/includes/class-wc-payments-translations-loader.php), the translations loader would no longer be necessary since from now on the JS file from `wp_register_script()` and the JSON hash would match.
I created #3493 to continue with that without delaying this fix.
